source/extensions/compression/zstd/common/dictionary_manager.h

```diff
@@ -0,0 +1,116 @@
+    bool has_watched = false;
+    watcher_ = dispatcher.createFilesystemWatcher();
+
+    DictionaryMapPtr dictionary_map(new DictionaryMap());
+    dictionary_map->reserve(dictionaries.size());
+
+    for (auto source : dictionaries) {
+      const auto data = Config::DataSource::read(source, false, api);
+      auto dictionary = DictionaryPtr(builder_(data.data(), data.length()));
+      auto id = getDictId(dictionary.get());
+      // If id == 0, the dictionary is not conformant to Zstandard specification, or empty.
+      if (id == 0) {
+        throw EnvoyException("Illegal Zstandard dictionary");
+      }
+      dictionary_map->emplace(id, std::move(dictionary));
+      if (source.specifier_case() ==
+          envoy::config::core::v3::DataSource::SpecifierCase::kFilename) {
+        has_watched = true;
+        const auto& filename = source.filename();
+        watcher_->addWatch(filename, Filesystem::Watcher::Events::Modified |
+                                         Filesystem::Watcher::Events::MovedTo,
+                           [this, id, filename](uint32_t) { onDictionaryUpdate(id, filename); });
+      }
+    }
+
+    tls_->set(
+        [dictionary_map](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr { return dictionary_map; });
+
+    if (!has_watched) {
+      watcher_.reset();
+    }
+  };
+
+  T* getDictionary(bool want_first, unsigned id) {
+    auto dictionary_map = tls_->getTyped<DictionaryMap>();
+
+    typename absl::flat_hash_map<unsigned, DictionaryPtr>::iterator it;
+    if (want_first) {
+      it = dictionary_map->begin();
+    } else {
+      it = dictionary_map->find(id);
+    }
+    if (it != dictionary_map->end()) {
+      return it->second.get();
+    }
+
+    return nullptr;
+  };
+
+  T* getDictionaryById(unsigned id) { return getDictionary(false, id); };
+
+  T* getFirstDictionary() { return getDictionary(true, 0); };
+
+private:
+  class DictionaryPtr : public std::shared_ptr<T> {
+  public:
+    DictionaryPtr(T* object) : std::shared_ptr<T>(object, deleter) {}
+  };
+  class DictionaryMap : public absl::flat_hash_map<unsigned, DictionaryPtr>, public ThreadLocal::ThreadLocalObject {
+  };
+  using DictionaryMapPtr = std::shared_ptr<DictionaryMap>;
+
+  void onDictionaryUpdate(unsigned origin_id, const std::string& filename) {
+    const auto data = api_.fileSystem().fileReadToEnd(filename);
+    if (!data.empty()) {
+      auto dictionary = DictionaryPtr(builder_(data.data(), data.length()));
+      auto id = getDictId(dictionary.get());
+      // Keep origin dictionary if the new is illegal
+      if (id != 0) {
+        tls_->runOnAllThreads(
+          [dictionary = std::move(dictionary), &tls = tls_]() {
+              auto dictionary_map = tls->getTyped<DictionaryMap>();
+              dictionary_map->emplace(id, dictionary);
+          });
+      }
+    }
+  }
+
+  Api::Api& api_;
+  ThreadLocal::SlotPtr tls_;
```

Comment:
It should be updated to `ThreadLocal::TypedSlotPtr<>` for better type-safe.

Reply:
Fix now